src/index.rs

```diff
@@ -231,8 +235,63 @@ impl IndexTable {
 		Ok(try_io!(Ok(&map[offset..offset + CHUNK_LEN])))
 	}
 
-	#[inline(never)]
 	fn find_entry(&self, key_prefix: u64, sub_index: usize, chunk: &[u8]) -> (Entry, usize) {
+		#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+		if is_x86_feature_detected!("sse2") {
+			return self.find_entry_sse2(key_prefix, sub_index, chunk)
+		}
+		self.find_entry_regular(key_prefix, sub_index, chunk)
+	}
+
+	#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+	fn find_entry_sse2(&self, key_prefix: u64, sub_index: usize, chunk: &[u8]) -> (Entry, usize) {
+		debug_assert!(
+			(CHUNK_ENTRIES - sub_index) % 4 == 0,
+			"SSE2 find entry only works on number of entities dividable by 4"
+		);
+
+		unsafe {
+			let partial_key = Entry::extract_key(key_prefix, self.id.index_bits());
+			let target = _mm_set1_epi32(partial_key as i32);
+			let shift = Entry::address_bits(self.id.index_bits());
+			let mut i = sub_index;
+			while i + 4 <= CHUNK_ENTRIES {
+				// We load the value two by two and do the same shift as for the partial_key
+				// before moving the values into the low part of the register
+				let first_two = _mm_shuffle_epi32::<0b11011000>(_mm_srl_epi64(
+					_mm_loadu_si128(chunk[i * 8..].as_ptr() as *const __m128i),
+					_mm_set_epi64x(0, shift.into()),
+				));
+				let last_two = _mm_shuffle_epi32::<0b11011000>(_mm_srl_epi64(
+					_mm_loadu_si128(chunk[(i + 2) * 8..].as_ptr() as *const __m128i),
+					_mm_set_epi64x(0, shift.into()),
+				));
+				// We set into current the low parts of first_two and last_two
+				// in the interleaved order
+				let current = _mm_unpacklo_epi32(first_two, last_two);
+				let cmp = _mm_movemask_epi8(_mm_cmpeq_epi32(current, target));
+				if cmp != 0 {
+					// We found a match, let's validate it
```

Comment:
I think here we can simply return entry at `i + cmp.leading_zeroes() / 4`

Reply:
The tricky thing is that the shift might be of 30 bits i.e. are left with a 34 bit number that we cast to 32 bits. So, this means we need to check again, sadly. But we can definitely be smarter and skip not useful comparison.

Reply:
Isn't the idea to just skip two bits of check in this case? (but then need to shift 32 bits instead of 30, so we check the 32 bit of the start, the 2 last unchecked bits will be checked with the rest of the key on value access).

Reply:
Spend time remembering/reunderstanding this today. might still be wrong, but my conclusion was the higher 2 so we in case of index 16 we would need to move >> 32 instead of 30. (the other way we would miss checking key content as the part of the key stored and check in index is 26 bytes containing the last two bits, then we indeed got 16 bit check from the access to the index, 32 from the index key content and 26 * 8 from the value table, should be 256).

Reply:
Thank you! I'm going to update the PR to use the high 32 bits.

Reply:
Done

Reply:
Previous version was actually more correct. We can't assume that partial key is always 32 bits, so shifts are still needed. We can ignore the last two bits for the cases when it is 33 or 34 bytes however. A bit of explanation:

Keys are always 256 bits. Key bits [48..256] are stored in the value table. So if `find_entry` returns a match, these bits are compared to make sure the rest of the key matches . Therefore `find_entry` must only match the first 48 bits of the key. Current version actually matches 50, so that's a bit of an overkill. 

The first `index_bits` of the key are simply the offset of the chunk in the file. All entries in the chunk share the same first `index_bits` of the key. They are not stored in the entries. Bits [index_bits..50] is what's actually stored in the entry and that's called `partial_key`.  The rest of the entry bits is `address`. As `index_bits` grows, so does `address_bits`, and `partial_key` shrinks. But the chunk index bits grows as well, so we can always match 50 bits of the key in the index file.

Indexes start with  `index_bits==16`. With  `index_bits==16` `partial_key` is 34, but the last two bits can be ignored and we only need to match 48 bits.
With `index_bits==21` however  `partial_key` is only 29 bits, so the last 3 of the first 32 entry bits should be excluded from comparison.

Reply:
Thank you for your detailed explanation. I indeed forgot that the index_bit can grow further. I have applied @cheme comments to fix the implementation.